kubernetes-membership-scheme/src/main/java/org/wso2/carbon/membership/scheme/kubernetes/resolver/ApiBasedPodIpResolver.java

```
@@ -0,0 +1,241 @@
+        final Set<String> containerIPs = new HashSet<>();
+
+        for (String kubernetesService : getKubernetesServicesArray()) {
+            // Create k8s api endpoint URL
+            URL apiEndpointUrl = createUrl(kubernetesApiServerUrl, apiContext + kubernetesService.trim());
+
+            // Create http/https k8s api endpoint
+            KubernetesApiEndpoint apiEndpoint = createAPIEndpoint(apiEndpointUrl);
+
+            // Create the connection and read k8s service endpoints
+            Endpoints endpoints;
+            try {
+                endpoints = getEndpoints(connectAndRead
+                        (apiEndpoint, kubernetesMasterUsername, kubernetesMasterPassword));
+
+            } catch (IOException e) {
+                throw new KubernetesMembershipSchemeException("Could not get the Endpoints", e);
+
+            } finally {
+                apiEndpoint.disconnect();
+            }
+
+            if (endpoints != null) {
+                if (endpoints.getSubsets() != null && !endpoints.getSubsets().isEmpty()) {
+                    // Reading IP addresses from two lists
+                    log.info("Reading IP addresses from endpoints");
+                    for (Subset subset : endpoints.getSubsets()) {
+                        if (subset.getAddresses() != null) {
+                            for (Address address : subset.getAddresses()) {
+                                containerIPs.add(address.getIp());
+                            }
+                        }
+                        if (subset.getNotReadyAddresses() != null) {
+                            for (Address address : subset.getNotReadyAddresses()) {
+                                containerIPs.add(address.getIp());
+                            }
+                        }
+                    }
+                }
+            } else {
+                log.info("No endpoints found at " + apiEndpointUrl.toString());
```

Thinking more, it would be better to throw an exception here yes.